src/Kestrel/KestrelConfigBuilder.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Server.Kestrel.Core;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal;
+using Microsoft.AspNetCore.Server.Kestrel.Https;
+using Microsoft.AspNetCore.Server.Kestrel.Https.Internal;
+using Microsoft.AspNetCore.Server.Kestrel.Internal;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Server.Kestrel
+{
+    public class KestrelConfigBuilder : IKestrelConfigBuilder
+    {
+        internal KestrelConfigBuilder(KestrelServerOptions options, IConfiguration configuration)
+        {
+            Options = options ?? throw new ArgumentNullException(nameof(options));
+            Configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+
+            Options.ConfigurationBuilder = this;
+        }
+
+        public KestrelServerOptions Options { get; }
+        public IConfiguration Configuration { get; }
+        private IDictionary<string, Action<EndpointConfiguration>> EndpointConfigurations { get; }
+            = new Dictionary<string, Action<EndpointConfiguration>>(0);
+
+        /// <summary>
+        /// Specifies a configuration Action to run when an endpoint with the given name is loaded from configuration.
+        /// </summary>
+        /// <param name="name"></param>
+        /// <param name="configureOptions"></param>
+        public KestrelConfigBuilder Endpoint(string name, Action<EndpointConfiguration> configureOptions)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            EndpointConfigurations[name] = configureOptions ?? throw new ArgumentNullException(nameof(configureOptions));
+            return this;
+        }
+
+        public void Build()
+        {
+            if (Options.ConfigurationBuilder == null)
+            {
+                // The builder has already been built.
```

Comment:
When does Build get called twice?

Reply:
I don't expect a given builder to be called twice, but it is a public API. If it was allowed to run twice you'd get duplicate endpoints and port conflicts.